### PR TITLE
Retries failover timeout

### PR DESCRIPTION
**Pull request by telent**
hey @dawid-sklodowski 

There's a PR to merge this into upstream moped but until we get an answer from them I'd like to get this fix into the branch we're using locally in Chopin

Not planning on releasing this live before Monday, btw

### CODE REVIEW

File: spec/moped/connection/socket/tcp_spec.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+describe Moped::Connection::Socket::TCP do
+  it "raises Moped::Errors::ConnectionFailure if no response within timeout" do
+    Timeout::timeout(10) do
+      expect {
+        # this test relies on the mongo protocol expecting the client
+        # to speak first
+        socket = described_class.connect('127.0.0.1', 27017, 2)
+        socket.read(100)
+      }.to raise_exception Moped::Errors::ConnectionFailure
```
Comment (dawid-sklodowski) [inline spec/moped/connection/socket/tcp_spec.rb:12]: Nice

File: lib/moped/connection/socket/connectable.rb [hunk 2]
```diff
@@ -44,7 +45,14 @@ def self.included(klass)
         # @since 1.2.0
         def read(length)
           check_if_alive!
-          handle_socket_errors { super }
+          handle_socket_errors {
+            if Kernel.select([self], nil, [self], @timeout)
+              super
+            else
```
Comment (dawid-sklodowski) [inline lib/moped/connection/socket/connectable.rb:51]: So before every read we will do select system call which will check whether socket is ready for read/write (with timeout).
Looks good.

**Comment by dawid-sklodowski** [comment]
signed-off-by: @dawid-sklodowski

Great job! Especially with low level system/protocol topics.

**Comment by joshuafleck** [comment]
Closing as this is 2 years old.

**Pull request closed without merging**